test/unit/specs/components/Identity/index.spec.js

```diff
@@ -59,5 +61,61 @@ describe("Identity", () => {
         expect(getEcid()).toBeUndefined();
       });
     });
+
+    describe("onResponse", () => {
+      let cookieJar;
+      let response;
+      beforeEach(() => {
+        cookieJar = {
+          get(key) {
+            return key === "ECID" ? "ABC" : null;
+          },
+          set: jasmine.createSpy()
+        };
+        response = {
+          getPayloadsByType: jasmine
+            .createSpy()
+            .and.returnValue([{ id: "ABC" }]),
+          toJSON: jasmine.createSpy()
+        };
+
+        identity = createIdentity({
+          config: {
+            reactorRegisterGetEcid
+          },
+          cookieJar
+        });
+        identity.lifecycle.onComponentsRegistered({
+          optIn: {
+            whenOptedIn() {
+              return Promise.resolve();
+            }
+          }
+        });
+      });
+      it("should get called with an object with a property named response in it", () => {
+        identity.lifecycle.onResponse({ response });
+        return flushPromiseChains().then(() => {
+          expect(response.getPayloadsByType).toHaveBeenCalledWith(
+            "identity:persist"
+          );
+          expect(cookieJar.set).toHaveBeenCalledWith(
+            EXPERIENCE_CLOUD_ID,
+            "ABC"
+          );
+          expect(response.getPayloadsByType).toHaveBeenCalledWith(
+            "identity:exchange"
```

Comment:
This is where having dependency injection makes things easy to test.  Because `idSyncs` is created inside the createIdentity function, there is no way to pass in a mock for it.  Therefore we cannot test that the result of response.getPayloadsByType("identity:exchange") is actually passed to idSyncs.process.  Ideally we would return a known value when the response is queried for "identity:exhange" and then just have verification here that idSyncs.merge was called with that value.

To follow dependency injection principles, you want to separate the code that creates objects with the code that does the business logic.  The code that creates the objects should be really simple, and will be tested by functional tests, but the code with the business logic should have unit tests.  In index.js, create all the objects you need, but then pass them into other functions or objects.  Then you can test the other functions or objects with mocks passed in.  You can look at Context/index.js for an example.

You are welcome to merge this PR without making this change, as this would be a fairly large refactor.

Reply:
I agree with this. Without restructuring the code, this test isn't nearly as valuable as it could it. By testing that `getPayloadsByType` is being called, we're getting a hint that something expected is happening, but we can't be confident that the information is being used correctly.